Upgrade `influxdb`

1.0 was released recently and contains breaking changes:

```
error[E0599]: no function or associated item named `write_query` found for trait object `dyn influxdb::query::Query` in the current scope
  --> src/metric_store/influxdb.rs:36:28
   |
36 |                     Query::write_query(Timestamp::Now, metric_owner.to_string());
   |                            ^^^^^^^^^^^ function or associated item not found in `dyn influxdb::query::Query`
```

--- a/rust/Cargo.toml
+++ b/rust/Cargo.toml
@@ -32,7 +32,8 @@ tracing-subscriber = "0.2.3"
 tracing-log = "0.1.1"
 tracing-futures = "0.2.3"
 
-influxdb = { version = "0.0.6", optional = true }
+influxdb = { version = "0.1.0", features = ["derive"], optional = true }
+chrono = { version = "0.4", features = ["serde"], optional = true }
 opentelemetry = { version = "0.2.0", optional = true }
 tracing-opentelemetry = { version = "0.2.0", optional = true }
 opentelemetry-jaeger = { version = "0.1.0", optional = true }
@@ -48,5 +49,5 @@ path = "src/bin/aggregator.rs"
 [features]
 default = []
 telemetry = [ "opentelemetry", "tracing-opentelemetry", "opentelemetry-jaeger"]
-influx_metrics = [ "influxdb" ]
+influx_metrics = [ "influxdb" , "chrono"]
 all = [ "telemetry", "influx_metrics"]

--- a/rust/src/bin/coordinator.rs
+++ b/rust/src/bin/coordinator.rs
@@ -9,7 +9,7 @@ use tracing_futures::Instrument;
 
 #[cfg(feature = "influx_metrics")]
 use xain_fl::{
-    common::metric_store::influxdb::{run_metricstore, InfluxDBMetricStore},
+    common::metric_store::influxdb::{run_metricstore, InfluxDBConnector},
     coordinator::settings::MetricStoreSettings,
 };
 
@@ -100,7 +100,7 @@ async fn _main(
     #[cfg(feature = "influx_metrics")]
     let metric_sender = if let Some(metric_store) = metric_store {
         // Start the metric store
-        let (influx_client, metric_sender) = InfluxDBMetricStore::new(
+        let (influx_client, metric_sender) = InfluxDBConnector::new(
             &metric_store.database_url[..],
             &metric_store.database_name[..],
         );

--- a/rust/src/common/metric_store/influxdb.rs
+++ b/rust/src/common/metric_store/influxdb.rs
@@ -71,7 +71,7 @@ pub async fn run_metricstore(mut influxdb_connector: InfluxDBConnector) {
                     Measurement::Round(round) => round.into_query("coordinator"),
                     Measurement::Counter(counter) => counter.into_query("coordinator"),
                 };
-                influxdb_connector
+                let _ = influxdb_connector
                     .client
                     .query(&query)
                     .await

--- a/rust/src/common/metric_store/influxdb.rs
+++ b/rust/src/common/metric_store/influxdb.rs
@@ -1,10 +1,19 @@
 use chrono::{DateTime, Utc};
-use influxdb::{Client, InfluxDbWriteable, Timestamp};
+use influxdb::{Client, InfluxDbWriteable, Timestamp, WriteQuery};
 use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
 
 pub enum Measurement {
     Round(RoundMeasurement),
     Counters(CountersMeasurement),
+}
+
+impl InfluxDbWriteable for Measurement {
+    fn into_query<I: Into<String>>(self, name: I) -> WriteQuery {
+        match self {
+            Self::Round(round) => round.into_query(name),
+            Self::Counters(counters) => counters.into_query(name),
+        }
+    }
 }
 
 #[derive(InfluxDbWriteable)]
@@ -67,13 +76,9 @@ pub async fn run_metricstore(mut influxdb_connector: InfluxDBConnector) {
     loop {
         match influxdb_connector.receiver.recv().await {
             Some(measurement) => {
-                let query = match measurement {
-                    Measurement::Round(round) => round.into_query("coordinator"),
-                    Measurement::Counters(counter) => counter.into_query("coordinator"),
-                };
                 let _ = influxdb_connector
                     .client
-                    .query(&query)
+                    .query(&measurement.into_query("coordinator"))
                     .await
                     .map_err(|e| error!("{}", e));
             }

--- a/rust/src/common/metric_store/influxdb.rs
+++ b/rust/src/common/metric_store/influxdb.rs
@@ -2,14 +2,34 @@ use chrono::{DateTime, Utc};
 use influxdb::{Client, InfluxDbWriteable, Timestamp};
 use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
 
+pub enum Measurement {
+    Round(RoundMeasurement),
+    Counter(CounterMeasurement),
+}
+
 #[derive(InfluxDbWriteable)]
-struct RoundMeasurement {
+pub struct RoundMeasurement {
     time: DateTime<Utc>,
     round: u32,
 }
 
+impl RoundMeasurement {
+    pub fn new(round: u32) -> RoundMeasurement {
+        RoundMeasurement {
+            time: Timestamp::Now.into(),
+            round,
+        }
+    }
+}
+
+impl From<RoundMeasurement> for Measurement {
+    fn from(value: RoundMeasurement) -> Self {
+        Self::Round(value)
+    }
+}
+
 #[derive(InfluxDbWriteable)]
-struct CounterMeasurement {
+pub struct CounterMeasurement {
     time: DateTime<Utc>,
     number_of_selected_participants: u32,
     number_of_waiting_participants: u32,
@@ -18,28 +38,50 @@ struct CounterMeasurement {
     number_of_ignored_participants: u32,
 }
 
-trait Metric {
-    fn to_measurement(&self) -> Box<dyn InfluxDbWriteable>;
+impl CounterMeasurement {
+    pub fn new(
+        number_of_selected_participants: u32,
+        number_of_waiting_participants: u32,
+        number_of_done_participants: u32,
+        number_of_done_inactive_participants: u32,
+        number_of_ignored_participants: u32,
+    ) -> CounterMeasurement {
+        CounterMeasurement {
+            time: Timestamp::Now.into(),
+            number_of_selected_participants,
+            number_of_waiting_participants,
+            number_of_done_participants,
+            number_of_done_inactive_participants,
+            number_of_ignored_participants,
+        }
+    }
 }
 
-pub async fn run_metricstore(mut influx_connector: InfluxDBConnector) {
+impl From<CounterMeasurement> for Measurement {
+    fn from(value: CounterMeasurement) -> Self {
+        Self::Counter(value)
+    }
+}
+
+pub async fn run_metricstore(mut influxdb_connector: InfluxDBConnector) {
     loop {
-        match influx_connector.receiver.recv().await {
-            Some(t) => {
-                // let mut write_query: WriteQuery =
-                //     Query::write_query(Timestamp::Now, metric_owner.to_string());
-
-                // for (name, value) in fields {
-                //     write_query = write_query.add_field(name, value);
-                // }
-
-                // // Submit the query to InfluxDB.
-                // let _ = influx_connector
-                //     .client
-                //     .query(&write_query)
-                //     .await
-                //     .map_err(|e| error!("{}", e));
-            }
+        match influxdb_connector.receiver.recv().await {
+            Some(measurement) => match measurement {
+                Measurement::Round(round) => {
+                    influxdb_connector
+                        .client
+                        .query(&round.into_query("coordinator"))
+                        .await
+                        .map_err(|e| eprintln!("{}", e));
+                }
+                Measurement::Counter(counter) => {
+                    influxdb_connector
+                        .client
+                        .query(&counter.into_query("coordinator"))
+                        .await
+                        .map_err(|e| eprintln!("{}", e));
+                }
+            },
             None => {
                 warn!("All senders have been dropped!");
                 return;
@@ -48,35 +90,13 @@ pub async fn run_metricstore(mut influx_connector: InfluxDBConnector) {
     }
 }
 
-// pub enum MetricOwner {
-//     Coordinator,
-//     Participant,
-// }
-
-// impl From<&MetricOwner> for &'static str {
-//     fn from(metric_owner: &MetricOwner) -> &'static str {
-//         match metric_owner {
-//             MetricOwner::Coordinator => "coordinator",
-//             MetricOwner::Participant => "participant",
-//         }
-//     }
-// }
-
-// impl ToString for MetricOwner {
-//     fn to_string(&self) -> String {
-//         Into::<&str>::into(self).into()
-//     }
-// }
-
-// pub struct Metric(pub MetricOwner, pub Vec<(&'static str, Type)>);
-
 pub struct InfluxDBConnector {
     client: Client,
-    receiver: UnboundedReceiver<Box<dyn Metric>>,
+    receiver: UnboundedReceiver<Measurement>,
 }
 
 impl InfluxDBConnector {
-    pub fn new(host: &str, db_name: &str) -> (InfluxDBConnector, UnboundedSender<Box<dyn Metric>>) {
+    pub fn new(host: &str, db_name: &str) -> (InfluxDBConnector, UnboundedSender<Measurement>) {
         let (sender, receiver) = unbounded_channel();
         (
             InfluxDBConnector {

--- a/rust/src/common/metric_store/influxdb.rs
+++ b/rust/src/common/metric_store/influxdb.rs
@@ -1,50 +1,44 @@
-use influxdb::{Client, Query, Timestamp, Type, WriteQuery};
+use chrono::{DateTime, Utc};
+use influxdb::{Client, InfluxDbWriteable, Timestamp};
 use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
 
-pub enum MetricOwner {
-    Coordinator,
-    Participant,
+#[derive(InfluxDbWriteable)]
+struct RoundMeasurement {
+    time: DateTime<Utc>,
+    round: u32,
 }
 
-impl From<&MetricOwner> for &'static str {
-    fn from(metric_owner: &MetricOwner) -> &'static str {
-        match metric_owner {
-            MetricOwner::Coordinator => "coordinator",
-            MetricOwner::Participant => "participant",
-        }
-    }
+#[derive(InfluxDbWriteable)]
+struct CounterMeasurement {
+    time: DateTime<Utc>,
+    number_of_selected_participants: u32,
+    number_of_waiting_participants: u32,
+    number_of_done_participants: u32,
+    number_of_done_inactive_participants: u32,
+    number_of_ignored_participants: u32,
 }
 
-impl ToString for MetricOwner {
-    fn to_string(&self) -> String {
-        Into::<&str>::into(self).into()
-    }
+trait Metric {
+    fn to_measurement(&self) -> Box<dyn InfluxDbWriteable>;
 }
 
-pub struct Metric(pub MetricOwner, pub Vec<(&'static str, Type)>);
-
-pub struct InfluxDBMetricStore {
-    client: Client,
-    receiver: UnboundedReceiver<Metric>,
-}
-
-pub async fn run_metricstore(mut influx_client: InfluxDBMetricStore) {
+pub async fn run_metricstore(mut influx_connector: InfluxDBConnector) {
     loop {
-        match influx_client.receiver.recv().await {
-            Some(Metric(metric_owner, fields)) => {
-                let mut write_query: WriteQuery =
-                    Query::write_query(Timestamp::Now, metric_owner.to_string());
+        match influx_connector.receiver.recv().await {
+            Some(t) => {
+                // let mut write_query: WriteQuery =
+                //     Query::write_query(Timestamp::Now, metric_owner.to_string());
 
-                for (name, value) in fields {
-                    write_query = write_query.add_field(name, value);
-                }
+                // for (name, value) in fields {
+                //     write_query = write_query.add_field(name, value);
+                // }
 
-                // Submit the query to InfluxDB.
-                let _ = influx_client
-                    .client
-                    .query(&write_query)
-                    .await
-                    .map_err(|e| error!("{}", e));
+                // // Submit the query to InfluxDB.
+                // let _ = influx_connector
+                //     .client
+                //     .query(&write_query)
+                //     .await
+                //     .map_err(|e| error!("{}", e));
             }
             None => {
                 warn!("All senders have been dropped!");
@@ -54,11 +48,38 @@ pub async fn run_metricstore(mut influx_client: InfluxDBMetricStore) {
     }
 }
 
-impl InfluxDBMetricStore {
-    pub fn new(host: &str, db_name: &str) -> (InfluxDBMetricStore, UnboundedSender<Metric>) {
+// pub enum MetricOwner {
+//     Coordinator,
+//     Participant,
+// }
+
+// impl From<&MetricOwner> for &'static str {
+//     fn from(metric_owner: &MetricOwner) -> &'static str {
+//         match metric_owner {
+//             MetricOwner::Coordinator => "coordinator",
+//             MetricOwner::Participant => "participant",
+//         }
+//     }
+// }
+
+// impl ToString for MetricOwner {
+//     fn to_string(&self) -> String {
+//         Into::<&str>::into(self).into()
+//     }
+// }
+
+// pub struct Metric(pub MetricOwner, pub Vec<(&'static str, Type)>);
+
+pub struct InfluxDBConnector {
+    client: Client,
+    receiver: UnboundedReceiver<Box<dyn Metric>>,
+}
+
+impl InfluxDBConnector {
+    pub fn new(host: &str, db_name: &str) -> (InfluxDBConnector, UnboundedSender<Box<dyn Metric>>) {
         let (sender, receiver) = unbounded_channel();
         (
-            InfluxDBMetricStore {
+            InfluxDBConnector {
                 client: Client::new(host, db_name),
                 receiver,
             },

--- a/rust/src/common/metric_store/influxdb.rs
+++ b/rust/src/common/metric_store/influxdb.rs
@@ -4,7 +4,7 @@ use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
 
 pub enum Measurement {
     Round(RoundMeasurement),
-    Counter(CounterMeasurement),
+    Counters(CountersMeasurement),
 }
 
 #[derive(InfluxDbWriteable)]
@@ -29,7 +29,7 @@ impl From<RoundMeasurement> for Measurement {
 }
 
 #[derive(InfluxDbWriteable)]
-pub struct CounterMeasurement {
+pub struct CountersMeasurement {
     time: DateTime<Utc>,
     number_of_selected_participants: u32,
     number_of_waiting_participants: u32,
@@ -38,15 +38,15 @@ pub struct CounterMeasurement {
     number_of_ignored_participants: u32,
 }
 
-impl CounterMeasurement {
+impl CountersMeasurement {
     pub fn new(
         number_of_selected_participants: u32,
         number_of_waiting_participants: u32,
         number_of_done_participants: u32,
         number_of_done_inactive_participants: u32,
         number_of_ignored_participants: u32,
-    ) -> CounterMeasurement {
-        CounterMeasurement {
+    ) -> CountersMeasurement {
+        CountersMeasurement {
             time: Timestamp::Now.into(),
             number_of_selected_participants,
             number_of_waiting_participants,
@@ -57,9 +57,9 @@ impl CounterMeasurement {
     }
 }
 
-impl From<CounterMeasurement> for Measurement {
-    fn from(value: CounterMeasurement) -> Self {
-        Self::Counter(value)
+impl From<CountersMeasurement> for Measurement {
+    fn from(value: CountersMeasurement) -> Self {
+        Self::Counters(value)
     }
 }
 
@@ -69,13 +69,13 @@ pub async fn run_metricstore(mut influxdb_connector: InfluxDBConnector) {
             Some(measurement) => {
                 let query = match measurement {
                     Measurement::Round(round) => round.into_query("coordinator"),
-                    Measurement::Counter(counter) => counter.into_query("coordinator"),
+                    Measurement::Counters(counter) => counter.into_query("coordinator"),
                 };
                 let _ = influxdb_connector
                     .client
                     .query(&query)
                     .await
-                    .map_err(|e| eprintln!("{}", e));
+                    .map_err(|e| error!("{}", e));
             }
             None => {
                 warn!("All senders have been dropped!");

--- a/rust/src/common/metric_store/influxdb.rs
+++ b/rust/src/common/metric_store/influxdb.rs
@@ -66,22 +66,17 @@ impl From<CounterMeasurement> for Measurement {
 pub async fn run_metricstore(mut influxdb_connector: InfluxDBConnector) {
     loop {
         match influxdb_connector.receiver.recv().await {
-            Some(measurement) => match measurement {
-                Measurement::Round(round) => {
-                    influxdb_connector
-                        .client
-                        .query(&round.into_query("coordinator"))
-                        .await
-                        .map_err(|e| eprintln!("{}", e));
-                }
-                Measurement::Counter(counter) => {
-                    influxdb_connector
-                        .client
-                        .query(&counter.into_query("coordinator"))
-                        .await
-                        .map_err(|e| eprintln!("{}", e));
-                }
-            },
+            Some(measurement) => {
+                let query = match measurement {
+                    Measurement::Round(round) => round.into_query("coordinator"),
+                    Measurement::Counter(counter) => counter.into_query("coordinator"),
+                };
+                influxdb_connector
+                    .client
+                    .query(&query)
+                    .await
+                    .map_err(|e| eprintln!("{}", e));
+            }
             None => {
                 warn!("All senders have been dropped!");
                 return;

--- a/rust/src/coordinator/core/service.rs
+++ b/rust/src/coordinator/core/service.rs
@@ -1,5 +1,5 @@
 #[cfg(feature = "influx_metrics")]
-use crate::common::metric_store::influxdb::{Metric, MetricOwner};
+use crate::common::metric_store::influxdb::{CounterMeasurement, Measurement, RoundMeasurement};
 use crate::{
     aggregator,
     common::client::{ClientId, Credentials, Token},
@@ -14,8 +14,6 @@ use crate::{
 };
 use derive_more::From;
 use futures::{ready, stream::Stream};
-#[cfg(feature = "influx_metrics")]
-use influxdb::Type;
 use std::{
     future::Future,
     pin::Pin,
@@ -100,7 +98,7 @@ where
 
     #[cfg(feature = "influx_metrics")]
     ///Metric Store
-    metrics_tx: Option<UnboundedSender<Metric>>,
+    metrics_tx: Option<UnboundedSender<Measurement>>,
 }
 
 impl<S> Service<S>
@@ -113,7 +111,7 @@ where
         aggregator_url: String,
         rpc_client: aggregator::rpc::Client,
         requests: ServiceRequests,
-        #[cfg(feature = "influx_metrics")] metrics_tx: Option<UnboundedSender<Metric>>,
+        #[cfg(feature = "influx_metrics")] metrics_tx: Option<UnboundedSender<Measurement>>,
     ) -> Self {
         let (heartbeat_expirations_tx, heartbeat_expirations_rx) = unbounded_channel();
 
@@ -416,41 +414,23 @@ where
     #[cfg(feature = "influx_metrics")]
     fn write_counter_metrics(&self) {
         self.metrics_tx.as_ref().map(|tx| {
-            let _ = tx.send(Metric(
-                MetricOwner::Coordinator,
-                vec![
-                    (
-                        "number_of_selected_participants",
-                        Type::SignedInteger(self.protocol.counters().selected as i64),
-                    ),
-                    (
-                        "number_of_waiting_participants",
-                        Type::SignedInteger(self.protocol.counters().waiting as i64),
-                    ),
-                    (
-                        "number_of_done_participants",
-                        Type::SignedInteger(self.protocol.counters().done as i64),
-                    ),
-                    (
-                        "number_of_done_inactive_participants",
-                        Type::SignedInteger(self.protocol.counters().done_and_inactive as i64),
-                    ),
-                    (
-                        "number_of_ignored_participants",
-                        Type::SignedInteger(self.protocol.counters().ignored as i64),
-                    ),
-                ],
-            ));
+            let _ = tx.send(
+                CounterMeasurement::new(
+                    self.protocol.counters().selected,
+                    self.protocol.counters().waiting,
+                    self.protocol.counters().done,
+                    self.protocol.counters().done_and_inactive,
+                    self.protocol.counters().ignored,
+                )
+                .into(),
+            );
         });
     }
 
     #[cfg(feature = "influx_metrics")]
     fn write_round_metric(&self, round: u32) {
         self.metrics_tx.as_ref().map(|tx| {
-            let _ = tx.send(Metric(
-                MetricOwner::Coordinator,
-                vec![("round", Type::UnsignedInteger(round as u64))],
-            ));
+            let _ = tx.send(RoundMeasurement::new(round).into());
         });
     }
 

--- a/rust/src/coordinator/core/service.rs
+++ b/rust/src/coordinator/core/service.rs
@@ -1,5 +1,5 @@
 #[cfg(feature = "influx_metrics")]
-use crate::common::metric_store::influxdb::{CounterMeasurement, Measurement, RoundMeasurement};
+use crate::common::metric_store::influxdb::{CountersMeasurement, Measurement, RoundMeasurement};
 use crate::{
     aggregator,
     common::client::{ClientId, Credentials, Token},
@@ -415,7 +415,7 @@ where
     fn write_counter_metrics(&self) {
         self.metrics_tx.as_ref().map(|tx| {
             let _ = tx.send(
-                CounterMeasurement::new(
+                CountersMeasurement::new(
                     self.protocol.counters().selected,
                     self.protocol.counters().waiting,
                     self.protocol.counters().done,

